### ISSUE TITLE
Liquibase failed when hyphen char in schema name

### ISSUE THREAD

**Issue by Bryksin** [liquibase/liquibase#1640]
Hi

As I understand according to [CORE-1371](https://liquibase.jira.com/browse/CORE-1371) this problem was fixed in liquibase 3.0.3

We are using the latest 4.2.2 however still getting a similar problem, here is the output log:

```
Starting Liquibase at 12:26:49 (version 4.2.2 #36 built at 2020-12-09 20:07+0000)
Unexpected error running Liquibase: liquibase.exception.DatabaseException: liquibase.exception.DatabaseException: 
Connection could not be created to jdbc:mysql://<***skipped-host***>:3360/timetable-solver-service_nspace25_INT_1329?createDatabaseIfNotExist=true with driver com.mysql.cj.jdbc.Driver.  
You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '-solver-service_nspace25_INT_1329' at line 1
For more information, please use the --logLevel flag
[2021-01-20 12:26:52] SEVERE [liquibase.integration] Unexpected error running Liquibase: liquibase.exception.DatabaseException: liquibase.exception.DatabaseException: Connection could not be created to jdbc:mysql://<***skipped-host***>.com:3360/timetable-solver-service_nspace25_INT_1329?createDatabaseIfNotExist=true with driver com.mysql.cj.jdbc.Driver.  You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '-solver-service_nspace25_INT_1329' at line 1
```

the schema name is defined by this pattern:
`schema-name: ${spring.application.name}_${ENV_SPACE}_${DB_SCHEMA_POSTFIX}`
where: 
`${spring.application.name}` = Service name
`${ENV_SPACE}` = Kubernetis space
`${DB_SCHEMA_POSTFIX}` = Git branch name

This approach was working for many services until we faced a case where service names contain a hyphen in the name.
We obviously can rename this parameter and replace all hyphens with underscores, but still... it should work with hyphens according to CORE-1371 fix

Any ideas?

**Comment by molivasdat** [comment]
Hi @Bryksin Thanks for writing up this issue.  The problem is not with liquibase.  In this case your jdbc string is invalid.
If you want to use - in your schema names, you are going to need to escape them with a \`
Since the single quote will be interpreted by the command line you will need to further escape the \` sign with a slash
So your jdbc url will need to look like this
```
\`schemas-with-a-hyphen-or-dash-symbol-in-the-name\`
```
In your case this will look like this
```
jdbc:mysql://<***skipped-host***>:3360/\`timetable-solver-service_nspace25_INT_1329\`?createDatabaseIfNotExist=true 
```
This link has further info
https://stackoverflow.com/questions/29275441/mysql-schema-name-with-dash-does-not-allow-me-to-execute-command-line-query
and this one
https://stackoverflow.com/questions/35592034/how-do-i-add-database-name-with-hyphen-character-using-script-in-ubuntu/35592181

I'm even having to escape my backtick with an escape character to write this info ;)

**Comment by Bryksin** [comment]
Hi

Thank you for your response, we added quotes as you suggested, still doesn't work

here is the output:
```
INFO: Executing:
docker run --rm --name liquibase-qa -v ~/.aws:/root/.aws -v /home/jenkins/workspace/t_qa-eks-app-deploy_INTINFRA-125/work/nspace25/liquibase/timetable-solver-service:/opt/liquibase/timetable-solver-service -e AWS_PROFILE=vsm-qa liquibase/liquibase:4.2 --url=jdbc:mysql://<***skipped-host***>:3360/\`timetable-solver-service_nspace25_INT_1329\`?createDatabaseIfNotExist=true --changeLogFile=changelog.xml --classpath=/opt/liquibase/timetable-solver-service --username=<skipped> --password=<skipped> --logLevel=warning update
Liquibase Community 4.2.2 by Datical
####################################################
##   _     _             _ _                      ##
##  | |   (_)           (_) |                     ##
##  | |    _  __ _ _   _ _| |__   __ _ ___  ___   ##
##  | |   | |/ _` | | | | | '_ \ / _` / __|/ _ \  ##
##  | |___| | (_| | |_| | | |_) | (_| \__ \  __/  ##
##  \_____/_|\__, |\__,_|_|_.__/ \__,_|___/\___|  ##
##              | |                               ##
##              |_|                               ##
##                                                ## 
##  Get documentation at docs.liquibase.com       ##
##  Get certified courses at learn.liquibase.com  ## 
##  Free schema change activity reports at        ##
##      https://hub.liquibase.com                 ##
##                                                ##
####################################################
Starting Liquibase at 12:50:48 (version 4.2.2 #36 built at 2020-12-09 20:07+0000)
Unexpected error running Liquibase: liquibase.exception.DatabaseException: liquibase.exception.DatabaseException: Connection could not be created to jdbc:mysql://<***skipped-host***>:3360/`timetable-solver-service_nspace25_INT_1329`?createDatabaseIfNotExist=true with driver com.mysql.cj.jdbc.Driver.  Unknown database '`timetable-solver-service_nspace25_INT_1329`'
```

**Comment by molivasdat** [comment]
@Bryksin Thanks for the feedback.  It looks like from the command line above you are using the liquibase docker image to run liquibase and are passing the url on the command line.  I was able to run with a locally installed liquibase command line with the parameters I outlined previously.  I will try to recreate with the docker image.  May need 2 escape characters like so `\\` before the \` because there are several command lines involved.  docker command line, the container command line and each one is interpreting the \`

**Comment by molivasdat** [comment]
Hi @Bryksin I was incorrect when I said I was able to connect from the command line.  I am not able to connect using jdbc connections at all 
Looks to be a bug with the database driver.
https://bugs.mysql.com/bug.php?id=95564
From all that I have read online, it seems like using a - in the database name in mysql may be more trouble that it is worth.

**Comment by ghoshbishakh** [comment]
I am facing the exact same issue @asashour

**Comment by mkavalecz** [comment]
@asashour @ghoshbishakh Try adding `<preserveSchemaCase>true</preserveSchemaCase>` to your `<configuration>` section as a workaround, until this gets fixed. It worked for me.

**Comment by vramik** [comment]
@tati-qalified IMO the workaround shouldn't be considered as permanent solution. 

I'm sure there are some users who needs the schema to be case insensitive and for those the workaround doesn't exist. Please consider re-opening the issue.

**Comment by mkavalecz** [comment]
Agreed. While it's perfectly fine as a workaround for some people, it's not a solution, and not even an universal workaround. And it should be an easy fix to make a proper solution.

**Comment by tati-qalified** [comment]
Thank you all for your feedback, I will be reopening this ticket. 
Would any of you be willing to submit a pull request for this?
Thanks!

**Comment by ahus1** [comment]
When looking at the code, it seems that the workaround with `preserveSchemaCase` suggested here only works for PostgreSQL, but looking at the code it will most likely not work for other databases.

I prepared a PR and would be happy to receive some feedback: #4897

**Comment by tati-qalified** [comment]
Hi all, I've re-tested this issue with Liquibase v4.31.0, MySQL and Postgres, Liquibase CLI and the Spring Boot integration, and couldn't replicate it.
@Bryksin @asashour @ghoshbishakh @vramik if any of you are still facing this problem with the latest Liquibase version, please leave a comment with more information and I'll reopen the issue.

**Comment by navi86** [comment]
@tati-qalified I got the same issue with Liquibase v4.31.0 Postgresql
`
[2025-04-07 08:30:08] INFO [liquibase.command] Logging exception.
ERROR: Exception Details
[2025-04-07 08:30:08] INFO [liquibase.ui] ERROR: Exception Details
ERROR: Exception Primary Class:  PSQLException
[2025-04-07 08:30:08] INFO [liquibase.ui] ERROR: Exception Primary Class:  PSQLException
ERROR: Exception Primary Reason:  ERROR: syntax error at or near "-"
  Position: 33
[2025-04-07 08:30:08] INFO [liquibase.ui] ERROR: Exception Primary Reason:  ERROR: syntax error at or near "-"
  Position: 33
ERROR: Exception Primary Source:  4.31.1
`
Do you have any suggestion how to fix it ?

**Comment by tati-qalified** [comment]
@navi86 have you tried using single or double quotes around the schema name?
